// @coffee-tree/sejong-auth-delegator/dist/esm/dto/ProfileResponseDto.js
class ProfileResponseDto {
    constructor(
        major,
        studentId,
        name,
        gradeLevel,
        userStatus,
        completedSemesters,
        verifiedSemesters
    ) {
        this.major = major;
        this.studentId = studentId;
        this.name = name;
        this.gradeLevel = this.extractNumber(gradeLevel);
        this.userStatus = userStatus;
        this.completedSemesters = this.extractNumber(completedSemesters);
        this.verifiedSemesters = this.extractNumber(verifiedSemesters);
    }

    extractNumber(value) {
        // 값이 존재하고 문자열인지를 확인
        if (value && typeof value === 'string') {
            return value.split(" ")[0];  // 예시로 "2학년"이 들어오면 "2"를 추출
        }
        return null;  // 값이 없거나 문자열이 아닌 경우
    }

    toObject() {
        return {
            major: this.major,
            studentId: this.studentCode,
            name: this.name,
            gradeLevel: this.gradeLevel,
            userStatus: this.userStatus,
            completedSemesters: this.completedSemesters,
            verifiedSemesters: this.verifiedSemesters,
        };
    }

    serialize() {
        return JSON.stringify(this.toObject());
    }
}

export default ProfileResponseDto;
