// @coffee-tree/sejong-auth-delegator/dist/esm/service/UserProfileService.js
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

class UserProfileService {
    constructor() {
        this.finalUrl = "https://classic.sejong.ac.kr/classic/reading/status.do";
    }

    async getUserProfile(ssotoken) {
        try {
            // 최종 데이터 요청
            const finalResponse = await fetch(this.finalUrl, {
                method: "GET",
                headers: {
                    "Cookie": `ssotoken=${ssotoken};`,
                },
                credentials: "include",
            });

            const html = await finalResponse.text();
            
            const profileData = this.extractProfileData(html); // 학과명, 학번, 이름, 학년 추출
            //console.log(profileData);
            return { body: profileData }; // response.body로 반환

        } catch (error) {
            console.error(error);
            throw new Error("Failed: Failed to fetch user profile");
        }
    }

    extractProfileData(html) {
        const $ = cheerio.load(html);

        // 필요한 데이터 추출
        const major = $("th:contains('학과명')").next().text().trim();
        const studentId = $("th:contains('학번')").next().text().trim();
        const name = $("th:contains('이름')").next().text().trim();
        const gradeLevel = $("th:contains('학년')").next().text().trim();

        if (!major || !studentId || !name || !gradeLevel) {
            throw new Error('필수 데이터가 누락되었습니다');
        }

        return {
            major,
            studentId,
            name,
            gradeLevel
        };
    }
}

export default UserProfileService;
