// @coffee-tree/sejong-auth-delegator/dist/esm/service/UserProfileService.js
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

class UserProfileService {
  constructor() {
    this.finalUrl = "https://classic.sejong.ac.kr/classic/reading/status.do";
  }

  async getUserProfile(ssotoken) {
    try {
      // ssotoken 쿠키를 사용해 요청
      const finalResponse = await fetch(this.finalUrl, {
        method: "GET",
        headers: {
          "Cookie": `ssotoken=${ssotoken};`,
        },
        credentials: "include",
      });

      const html = await finalResponse.text();

      // html에서 프로필 데이터 추출
      const profileData = this.extractProfileData(html);
      return { body: profileData };

    } catch (error) {
      console.error(error);
      throw new Error("Failed: Failed to fetch user profile");
    }
  }

  extractProfileData(html) {
    const $ = cheerio.load(html);

    // 원하는 데이터 추출
    const major = $("th:contains('학과명')").next().text().trim();
    const studentId = $("th:contains('학번')").next().text().trim();
    const name = $("th:contains('이름')").next().text().trim();
    const gradeLevel = $("th:contains('학년')").next().text().trim();

    if (!major || !studentId || !name || !gradeLevel) {
      throw new Error('필수 데이터가 누락되었습니다');
    }

    return {
      major,
      studentId,
      name,
      gradeLevel
    };
  }
}

export default UserProfileService;
